primitives/runtime/src/offchain/rpc.rs

```diff
@@ -0,0 +1,435 @@
+	/// rpc request jsonrpc
+	pub jsonrpc: &'a str,
+	/// rpc request id
+	pub id: u32,
+	/// rpc request method
+	pub method: &'a str,
+	/// rpc request params
+	pub params: Vec<&'a str>,
+}
+
+/// A rpc client
+#[derive(Clone, RuntimeDebug)]
+pub struct Client<'a> {
+	/// RPC request timeout in milliseconds.
+	pub timeout: u64,
+	/// HTTP URL to send the RPC request to.
+	pub url: &'a str,
+}
+
+impl<'a> Default for Request<'a> {
+	fn default() -> Self {
+		Request {
+			jsonrpc: "2.0",
+			id: 1,
+			method: "chain_getFinalizedHead",
+			params: Vec::new(),
+		}
+	}
+}
+
+impl<'a> Default for Client<'a> {
+	fn default() -> Self {
+		Client {
+			timeout: 3_000,
+			url: "http://localhost:9933"
+		}
+	}
+}
+
+impl<'a> Request<'a> {
+	/// Create a new Request
+	pub fn new() -> Self {
+		Request::default()
+	}
+
+	/// Change the Method for the Request
+	pub fn method(mut self, method: &'a str) -> Self {
+		self.method = method;
+		self
+	}
+
+	/// Change the Params for the Request
+	pub fn params(mut self, params: Vec<&'a str>) -> Self {
+		self.params = params;
+		self
+	}
+
+	/// Change the JSON-RPC version of the Request
+	pub fn jsonrpc(mut self, jsonrpc: &'a str) -> Self {
+		self.jsonrpc = jsonrpc;
+		self
+	}
+
+	/// Change the ID for the Request
+	pub fn id(mut self, id: u32) -> Self {
+		self.id = id;
+		self
+	}
+}
+
+impl<'a> Client<'a> {
+	/// Create a new Client
+	pub fn new() -> Self {
+		Client::default()
+	}
+
+	/// Change the URL for the Client
+	pub fn url(mut self, url: &'a str) -> Self {
+		self.url = url;
+		self
+	}
+
+	/// Change the Timeout for the Client
+	pub fn timeout(mut self, timeout: u64) -> Self {
+		self.timeout = timeout;
+		self
+	}
+
+	/// Send the request with this Client's `timeout` and return a RPC result.
+	///
+	/// `Err` is returned in case of underlying HTTP error or deserialization error.
+	pub fn send(&self, request: &'a Request) -> RpcResult {
+		let deadline = sp_io::offchain::timestamp().add(Duration::from_millis(self.timeout));
+		self.send_with_deadline(deadline, request)
+	}
+
+	/// Send the request with a completion deadline and return a RPC result.
+	///
+	/// Note that unlike [`send`], given `deadline` is used instead of the default Client's `timeout`.
+	/// `Err` is returned in case of underlying HTTP error or deserialization error.
+	pub fn send_with_deadline(&self, deadline: Timestamp, request: &'a Request) -> RpcResult {
+		// Construct http POST body.
+		// Note that we explicitly avoid deriving `Serialize` from `serde`
+		// , cause it significantly increases the runtime size.
+		let request_body = json!({
+			"jsonrpc": request.jsonrpc,
+			"id": request.id,
+			"method": request.method,
+			"params": request.params
+		});
+		let mut body: Vec<&[u8]> = Vec::new();
+		let request_body_slice: &[u8] = &(serde_json::to_vec(&request_body)
+			.expect("Constructed request body has infallible serialization, just an object with strings; qed"))[..];
+		body.push(request_body_slice);
```

Comment:
```suggestion
		let body = vec![
			serde_json::to_vec(&request_body)
				.expect("Constructed request body has infallible serialization, just an object with strings; qed")
		];
```
This should work as well, but is a bit simpler.